Wiki Collaborators

thomasjacobjr opened the pull request:
-- ran migration

thomasjacobjr committed:
Created Collab migration file

-- ran migration

thomasjacobjr committed:
Created Collab Model

thomasjacobjr committed:
Renamed Collaborators to Collaborations

thomasjacobjr committed:
Updated User Model

thomasjacobjr committed:
Refactored user, wiki, collab models

thomasjacobjr committed:
Final refactor of model delegations

thomasjacobjr committed:
Updated wiki policy scope

thomasjacobjr committed:
Added default value for Private Wikis to False

- Ran migration and reset DB

thomasjacobjr committed:
Authorized anonymous users for home page

thomasjacobjr committed:
Defined public method for Wiki class

thomasjacobjr committed:
WIP fixing wiki policy scope

thomasjacobjr committed:
Refactored Wiki Policy Scope for Users

Refactored wiki_policy.rb to properly filter roles / wikis for index and show
views

thomasjacobjr committed:
WIP - Refactoring wiki and user model associations

chikamichi commented on app/controllers/wikis_controller.rb:
```diff
@@ -25,7 +26,11 @@ def create
+
+    if params[:wiki][:private]
+      @wiki.private = params[:wiki][:private]
+    end
```
Just FYI, a slightly better, cleaner way of checking for the presence of the parameter would be:

``` ruby
unless params[:wiki][:private].nil?
  @wiki.private = params[:wiki][:private]
end
```

Checking for "nilness" allows for `false` to be still considered an actual value for the parameter. Right now it would be discarded, which is _fine_ because the default value of the private attribute is `false` anyway, but…

chikamichi committed:
Fixing aliases on Wiki/User/Collaborations HMT

Has Many Through (HTM) associations can use aliases, as long as the
names of the joined models is provided with the :source option.

Fixed the improper "private_wiki" model name in the join model,
Collaboration, to "wiki".

chikamichi commented on app/models/collaboration.rb:
```diff
@@ -1,4 +1,4 @@
   belongs_to :user
-  belongs_to :private_wiki 
+  belongs_to :wiki
 end
```
This error I overlooked. `Collaboration` must reference the actual model names, `User` (as `:user`) and `Wiki` (as `:wiki`). The `private_wiki` name is defined in app/models/user.rb as an alias (see below).

Remember, the `collaborations` table has two columns named `user_id` and `wiki_id`, so the `belongs_to` declarations must match those.

chikamichi commented on app/models/user.rb:
```diff
@@ -1,7 +1,8 @@
   has_many :collaborations
-  has_many :private_wikis, through: :collaborations
+  has_many :private_wikis, through: :collaborations, source: :wiki
+  
```
`Collaboration` only knows about wikis (`belongs_to :wiki`), so this line actually defines an alias, `private_wikis`, for a collaboration's wikis attribute.

It basically means that calling `some_user.private_wikis` will go look in the `collaborations` table and fetch all the records where the `user_id` column has a value of `some_user.id`; this list of Collaboration records will be mapped to a list of `Wiki` ids (`wiki_id` column); then the matching wikis will be fetched from the `wikis` table and returned to you.

chikamichi replied:
Quick remark: `private_wikis` may not be that good of a name, because a user could totally be the _owner_ of some private wikis, and one could expect `some_user.private_wikis` to return that list of owned, private wikis, instead of a list of invited-to, private wikis.

One could simply rename the HMT to something more explicit then:

``` ruby
has_many :wiki_invites, through: :collaborations, source: :wiki
```

``` ruby
some_user.wiki_invites # returns a list of private wikis the user is a collaborator on
```

chikamichi commented on app/models/wiki.rb:
```diff
@@ -1,7 +1,7 @@
   has_many :collaborations
-  has_many :users, through: :collaborations
+  has_many :collaborators, through: :collaborations, source: :user
 
```
Same here, `Collaboration` only knows about users (`belongs_to :user`), so this line defines an alias, `collaborators`, for a collaboration's users attribute.

I really like this line, because it says: "In the context of a collaboration, a private wiki may be bound to specific users known as collaborators", which makes total sense.

It basically means that calling `some_wiki.collaborators` will go look in the `collaborations` table and fetch all the records where the `wiki_id` column has a value of `wiki.id`; this list of `Collaboration` records will be mapped to a list of `User` ids (`user_id` column); then the matching users will be fetched from the `users` table and returned to you.

chikamichi commented:
With this latest commit, you should now be able to do stuff like this in the rails console:

``` ruby
wiki = Wiki.last
wiki.collaborators # returns an empty list (of User instances)

wiki.collaborators << User.first
wiki.collaborators # returns a list that is not empty anymore!

wiki.collaborations # returns a list of Collaboration instances, right now one item long
# Note: it is unlikely you will ever access wiki.collaborations and do stuff with it.
# Just to showcase it is there, persisting the relationship between the user and the wiki.

dude = wiki.collaborators.first
dude.private_wikis # returns a list of Wiki instances the dude is a collaborator on
```

Give it a try, check the returned values as you are doing those operations!

---

If you have some pending edits locally, run `git stash` before `git pull`, then `git stash pop` if you want to reapply your stashed edits.

chikamichi commented:
Hint: if you want to more easily look at the tables themselves, you may install [this gem](https://github.com/igorkasyanchuk/rails_db) (scroll down to the _Install_ section). It is pretty straightforward and could make your life easier, alongside using the Rails console.

thomasjacobjr committed:
Refactored HMT associations in models

thomasjacobjr committed:
Merge remote wiki-collaborators into local wiki-collaborators

thomasjacobjr committed:
Installed Rails DB Gem

thomasjacobjr committed:
Added Collab text field to Edit Wiki view

thomasjacobjr committed:
Added ability to add collaborators
for wikis

thomasjacobjr committed:
Display error message if unable to add collaborator emails

thomasjacobjr committed:
Let user add collaborators to new wikis

thomasjacobjr committed:
Standard users can edit private wikis if they are collabs.

thomasjacobjr committed:
No more error displayed if email collab is empty.

thomasjacobjr committed:
Prevent duplicate collaborators on wikis.

Refactored logic into its own shared method,
add_collaborators

thomasjacobjr committed:
User can view collaborators for a wiki.

thomasjacobjr committed:
User can remove collaborators.

Added Collaborators Controller and Destory action

thomasjacobjr committed:
Check authorization upon removing collabs

TO FIX - A standard user collab can't remove self from list.

chikamichi commented:
😻
